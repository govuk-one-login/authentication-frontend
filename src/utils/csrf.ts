import { csrfSync } from "csrf-sync";

const { csrfSynchronisedProtection } = csrfSync({
  getTokenFromRequest: (req) => {
    return req.body["_csrf"];
  },
});

export { csrfSynchronisedProtection };
