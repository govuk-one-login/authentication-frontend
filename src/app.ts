import express from "express";
import cookieParser from "cookie-parser";
import csurf from "csurf";
import { loggerMiddleware } from "./utils/logger";

import { sanitizeRequestMiddleware } from "./middleware/sanitize-request-middleware";
import i18nextMiddleware from "i18next-http-middleware";
import * as path from "path";
import { configureNunjucks } from "./config/nunchucks";
import { i18nextConfigurationOptions } from "./config/i18next";
import { helmetConfiguration } from "./config/helmet";
import helmet from "helmet";

import { setHtmlLangMiddleware } from "./middleware/html-lang-middleware";
import i18next from "i18next";
import Backend from "i18next-fs-backend";

import {
  getAppEnv,
  getNodeEnv,
  getRedisConfig,
  getRedisHost,
  getRedisPort,
  getSessionExpiry,
  getSessionSecret,
  supportAccountRecovery,
} from "./config";
import { logErrorMiddleware } from "./middleware/log-error-middleware";
import { getCookieLanguageMiddleware } from "./middleware/cookie-lang-middleware";
import { enterEmailRouter } from "./components/enter-email/enter-email-routes";
import { enterPasswordRouter } from "./components/enter-password/enter-password-routes";
import { footerRouter } from "./components/common/footer/footer-pages-routes";
import { registerAccountCreatedRouter } from "./components/account-created/account-created-routes";
import { createPasswordRouter } from "./components/create-password/create-password-routes";
import { enterPhoneNumberRouter } from "./components/enter-phone-number/enter-phone-number-routes";

import { pageNotFoundHandler } from "./handlers/page-not-found-handler";
import { serverErrorHandler } from "./handlers/internal-server-error-handler";
import { csrfMiddleware } from "./middleware/csrf-middleware";
import { checkYourPhoneRouter } from "./components/check-your-phone/check-your-phone-routes";
import { landingRouter } from "./components/landing/landing-route";
import { getCSRFCookieOptions } from "./config/cookie";
import { ENVIRONMENT_NAME } from "./app.constants";
import { enterMfaRouter } from "./components/enter-mfa/enter-mfa-routes";
import { authCodeRouter } from "./components/auth-code/auth-code-routes";
import { resendMfaCodeRouter } from "./components/resend-mfa-code/resend-mfa-code-routes";
import { resendEmailCodeRouter } from "./components/resend-email-code/resend-email-code-routes";
import { signedOutRouter } from "./components/signed-out/signed-out-routes";
import {
  getSessionIdMiddleware,
  initialiseSessionMiddleware,
} from "./middleware/session-middleware";
import { shareInfoRouter } from "./components/share-info/share-info-routes";
import { updatedTermsConditionsRouter } from "./components/updated-terms-conditions/updated-terms-conditions-routes";
import { signInOrCreateRouter } from "./components/sign-in-or-create/sign-in-or-create-routes";
import { accountNotFoundRouter } from "./components/account-not-found/account-not-found-routes";
import { resetPasswordCheckEmailRouter } from "./components/reset-password-check-email/reset-password-check-email-routes";
import { setLocalVarsMiddleware } from "./middleware/set-local-vars-middleware";
import { resetPasswordRouter } from "./components/reset-password/reset-password-routes";
import { noCacheMiddleware } from "./middleware/no-cache-middleware";
import { checkYourEmailRouter } from "./components/check-your-email/check-your-email-routes";
import { securityCodeErrorRouter } from "./components/security-code-error/security-code-error-routes";
import { upliftJourneyRouter } from "./components/uplift-journey/uplift-journey-routes";
import { contactUsRouter } from "./components/contact-us/contact-us-routes";
import { getSessionCookieOptions, getSessionStore } from "./config/session";
import session from "express-session";
import { proveIdentityRouter } from "./components/prove-identity/prove-identity-routes";
import { healthcheckRouter } from "./components/healthcheck/healthcheck-routes";
import { crossDomainTrackingMiddleware } from "./middleware/cross-domain-tracking-middleware";
import { proveIdentityWelcomeRouter } from "./components/prove-identity-welcome/prove-identity-welcome-routes";
import { proveIdentityCallbackRouter } from "./components/prove-identity-callback/prove-identity-callback-routes";
import { docCheckingAppRouter } from "./components/doc-checking-app/doc-checking-app-routes";
import { docCheckingAppCallbackRouter } from "./components/doc-checking-app-callback/doc-checking-app-callback-routes";
import { selectMFAOptionsRouter } from "./components/select-mfa-options/select-mfa-options-routes";
import { setupAuthenticatorAppRouter } from "./components/setup-authenticator-app/setup-authenticator-app-routes";
import { enterAuthenticatorAppCodeRouter } from "./components/enter-authenticator-app-code/enter-authenticator-app-code-routes";
import { cookiesRouter } from "./components/common/cookies/cookies-routes";
import { errorPageRouter } from "./components/common/errors/error-routes";
import { setInternationalPhoneNumberSupportMiddleware } from "./middleware/set-international-phone-number-support-middleware";
import { cannotChangeSecurityCodesRouter } from "./components/cannot-change-security-codes/cannot-change-security-codes-routes";
import { checkYourEmailSecurityCodesRouter } from "./components/check-your-email-security-codes/check-your-email-security-codes-routes";

const APP_VIEWS = [
  path.join(__dirname, "components"),
  path.resolve("node_modules/govuk-frontend/"),
];

function registerRoutes(app: express.Application) {
  app.use(landingRouter);
  app.use(signInOrCreateRouter);
  app.use(enterEmailRouter);
  app.use(accountNotFoundRouter);
  app.use(enterPasswordRouter);
  app.use(resetPasswordCheckEmailRouter);
  app.use(checkYourEmailRouter);
  app.use(createPasswordRouter);
  app.use(selectMFAOptionsRouter);
  app.use(enterAuthenticatorAppCodeRouter);
  app.use(setupAuthenticatorAppRouter);
  app.use(enterPhoneNumberRouter);
  app.use(registerAccountCreatedRouter);
  app.use(footerRouter);
  app.use(checkYourPhoneRouter);
  if (supportAccountRecovery()) {
    app.use(cannotChangeSecurityCodesRouter);
    app.use(checkYourEmailSecurityCodesRouter);
  }
  app.use(securityCodeErrorRouter);
  app.use(enterMfaRouter);
  app.use(authCodeRouter);
  app.use(resendMfaCodeRouter);
  app.use(resendEmailCodeRouter);
  app.use(signedOutRouter);
  app.use(shareInfoRouter);
  app.use(updatedTermsConditionsRouter);
  app.use(resetPasswordRouter);
  app.use(upliftJourneyRouter);
  app.use(contactUsRouter);
  app.use(healthcheckRouter);
  app.use(proveIdentityRouter);
  app.use(proveIdentityWelcomeRouter);
  app.use(proveIdentityCallbackRouter);
  app.use(cookiesRouter);
  app.use(docCheckingAppRouter);
  app.use(docCheckingAppCallbackRouter);
  app.use(errorPageRouter);
}

async function createApp(): Promise<express.Application> {
  const app: express.Application = express();
  const isProduction = getNodeEnv() === ENVIRONMENT_NAME.PROD;

  app.enable("trust proxy");

  app.use(express.json());
  app.use(express.urlencoded({ extended: false }));
  app.use(noCacheMiddleware);
  app.use(loggerMiddleware);

  app.use(
    "/assets",
    express.static(path.resolve("node_modules/govuk-frontend/govuk/assets"))
  );

  app.use("/public", express.static(path.join(__dirname, "public")));
  app.set("view engine", configureNunjucks(app, APP_VIEWS));
  app.use(setLocalVarsMiddleware);

  i18next
    .use(Backend)
    .use(i18nextMiddleware.LanguageDetector)
    .init(
      i18nextConfigurationOptions(
        path.join(__dirname, "locales/{{lng}}/{{ns}}.json")
      )
    );

  app.use(i18nextMiddleware.handle(i18next));
  app.use(helmet(helmetConfiguration));

  const redisConfig = isProduction
    ? await getRedisConfig(getAppEnv())
    : { host: getRedisHost(), port: getRedisPort(), isLocal: true };

  app.use(
    session({
      name: "aps",
      store: getSessionStore(redisConfig),
      saveUninitialized: false,
      secret: getSessionSecret(),
      unset: "destroy",
      resave: false,
      cookie: getSessionCookieOptions(
        isProduction,
        getSessionExpiry(),
        getSessionSecret()
      ),
    })
  );

  app.use(cookieParser());
  app.use(csurf({ cookie: getCSRFCookieOptions(isProduction) }));

  app.use(getSessionIdMiddleware);
  app.post("*", sanitizeRequestMiddleware);
  app.use(csrfMiddleware);
  app.use(setHtmlLangMiddleware);
  app.use(initialiseSessionMiddleware);
  app.use(crossDomainTrackingMiddleware);
  app.use(setInternationalPhoneNumberSupportMiddleware);

  registerRoutes(app);

  app.use(getCookieLanguageMiddleware);

  app.use(logErrorMiddleware);
  app.use(serverErrorHandler);
  app.use(pageNotFoundHandler);

  return app;
}

export { createApp };
