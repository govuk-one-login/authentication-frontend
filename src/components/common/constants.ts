import { PATH_NAMES } from "../../app.constants";
import { getNextState } from "./state-machine/state-machine";

export const SECURITY_CODE_ERROR = "actionType";

export enum SecurityCodeErrorType {
  MfaMaxCodesSent = "mfaMaxCodesSent",
  MfaBlocked = "mfaBlocked",
  MfaMaxRetries = "mfaMaxRetries",
  OtpMaxCodesSent = "otpMaxCodesSent",
  OtpBlocked = "otpBlocked",
  OtpMaxRetries = "otpMaxRetries",
  EmailMaxCodesSent = "emailMaxCodesSent",
  EmailBlocked = "emailBlocked",
  EmailMaxRetries = "emailMaxRetries",
  AuthAppMfaNoCodeValidator = "authAppMfaNoCodeValidator",
  AuthAppMfaMaxRetries = "authAppMfaMaxRetries",
  ChangeSecurityCodesEmailMaxCodesSent = "changeSecurityCodesEmailMaxCodesSent",
  ChangeSecurityCodesEmailBlocked = "changeSecurityCodesEmailBlocked",
  ChangeSecurityCodesEmailMaxRetries = "changeSecurityCodesEmailMaxRetries",
  InvalidPasswordResetCodeMaxRetries = "invalidPasswordResetCodeMaxRetries",
}

export const ERROR_CODES = {
  SESSION_ID_MISSING_OR_INVALID: 1000,
  RESET_PASSWORD_LINK_MAX_RETRIES_REACHED: 1022,
  RESET_PASSWORD_LINK_BLOCKED: 1023,
  NEW_PASSWORD_SAME_AS_EXISTING: 1024,
  ENTERED_INVALID_MFA_MAX_TIMES: 1027,
  VERIFY_EMAIL_CODE_REQUEST_BLOCKED: 1029,
  VERIFY_PHONE_NUMBER_CODE_REQUEST_BLOCKED: 1030,
  VERIFY_PHONE_NUMBER_MAX_CODES_SENT: 1032,
  VERIFY_EMAIL_MAX_CODES_SENT: 1031,
  ENTERED_INVALID_VERIFY_EMAIL_CODE_MAX_TIMES: 1033,
  ENTERED_INVALID_VERIFY_PHONE_NUMBER_CODE_MAX_TIMES: 1034,
  MFA_SMS_MAX_CODES_SENT: 1025,
  MFA_CODE_REQUESTS_BLOCKED: 1026,
  INVALID_MFA_CODE: 1035,
  INVALID_VERIFY_EMAIL_CODE: 1036,
  INVALID_VERIFY_PHONE_NUMBER_CODE: 1037,
  ENTERED_INVALID_PASSWORD_RESET_CODE_MAX_TIMES: 1039,
  INVALID_PASSWORD_MAX_ATTEMPTS_REACHED: 1028,
  RESET_PASSWORD_INVALID_CODE: 1021,
  PASSWORD_IS_COMMON: 1040,
  AUTH_APP_INVALID_CODE_MAX_ATTEMPTS_REACHED: 1042,
  AUTH_APP_INVALID_CODE: 1043,
  ACCOUNT_LOCKED: 1045,
  VERIFY_CHANGE_HOW_GET_SECURITY_CODES_MAX_CODES_SENT: 1046,
  VERIFY_CHANGE_HOW_GET_SECURITY_CODES_CODE_REQUEST_BLOCKED: 1047,
  VERIFY_CHANGE_HOW_GET_SECURITY_CODES_INVALID_CODE: 1048,
};

export const ERROR_CODE_MAPPING: { [p: string]: string } = {
  [ERROR_CODES.ACCOUNT_LOCKED]: pathWithQueryParam(
    PATH_NAMES["ACCOUNT_LOCKED"]
  ),
  [ERROR_CODES.INVALID_PASSWORD_MAX_ATTEMPTS_REACHED]: pathWithQueryParam(
    PATH_NAMES["ACCOUNT_LOCKED"]
  ),
  [ERROR_CODES.MFA_SMS_MAX_CODES_SENT]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_REQUEST_EXCEEDED"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.MfaMaxCodesSent
  ),
  [ERROR_CODES.MFA_CODE_REQUESTS_BLOCKED]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_WAIT"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.MfaBlocked
  ),
  [ERROR_CODES.ENTERED_INVALID_MFA_MAX_TIMES]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_INVALID"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.MfaMaxRetries
  ),
  [ERROR_CODES.VERIFY_PHONE_NUMBER_MAX_CODES_SENT]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_REQUEST_EXCEEDED"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.OtpMaxCodesSent
  ),
  [ERROR_CODES.VERIFY_PHONE_NUMBER_CODE_REQUEST_BLOCKED]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_WAIT"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.OtpBlocked
  ),
  [ERROR_CODES.ENTERED_INVALID_VERIFY_PHONE_NUMBER_CODE_MAX_TIMES]:
    pathWithQueryParam(
      PATH_NAMES["SECURITY_CODE_INVALID"],
      SECURITY_CODE_ERROR,
      SecurityCodeErrorType.OtpMaxRetries
    ),
  [ERROR_CODES.VERIFY_EMAIL_CODE_REQUEST_BLOCKED]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_REQUEST_EXCEEDED"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.EmailBlocked
  ),
  [ERROR_CODES.ENTERED_INVALID_VERIFY_EMAIL_CODE_MAX_TIMES]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_INVALID"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.EmailMaxRetries
  ),
  [ERROR_CODES.ENTERED_INVALID_PASSWORD_RESET_CODE_MAX_TIMES]:
    pathWithQueryParam(
      PATH_NAMES["SECURITY_CODE_INVALID"],
      SECURITY_CODE_ERROR,
      SecurityCodeErrorType.InvalidPasswordResetCodeMaxRetries
    ),
  [ERROR_CODES.VERIFY_EMAIL_MAX_CODES_SENT]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_WAIT"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.EmailMaxCodesSent
  ),
  [ERROR_CODES.AUTH_APP_INVALID_CODE_MAX_ATTEMPTS_REACHED]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_INVALID"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.AuthAppMfaMaxRetries
  ),
  [ERROR_CODES.VERIFY_CHANGE_HOW_GET_SECURITY_CODES_MAX_CODES_SENT]:
    pathWithQueryParam(
      PATH_NAMES["SECURITY_CODE_REQUEST_EXCEEDED"],
      SECURITY_CODE_ERROR,
      SecurityCodeErrorType.ChangeSecurityCodesEmailMaxCodesSent
    ),
  [ERROR_CODES.VERIFY_CHANGE_HOW_GET_SECURITY_CODES_INVALID_CODE]:
    pathWithQueryParam(
      PATH_NAMES["SECURITY_CODE_INVALID"],
      SECURITY_CODE_ERROR,
      SecurityCodeErrorType.ChangeSecurityCodesEmailMaxRetries
    ),
  [ERROR_CODES.VERIFY_CHANGE_HOW_GET_SECURITY_CODES_CODE_REQUEST_BLOCKED]:
    pathWithQueryParam(
      PATH_NAMES["SECURITY_CODE_WAIT"],
      SECURITY_CODE_ERROR,
      SecurityCodeErrorType.ChangeSecurityCodesEmailBlocked
    ),
};

export function getErrorPathByCode(errorCode: number): string | undefined {
  const nextPath = ERROR_CODE_MAPPING[errorCode.toString()];

  if (!nextPath) {
    return undefined;
  }

  return nextPath;
}

export function pathWithQueryParam(
  path: string,
  queryParam?: string,
  value?: string | SecurityCodeErrorType
): string {
  if (queryParam && value) {
    const queryParams = new URLSearchParams({
      [queryParam]: value,
    }).toString();

    return path + "?" + queryParams;
  }

  return path;
}

export function getNextPathAndUpdateJourney(
  req: any,
  path: string,
  event: string,
  ctx?: any,
  sessionId?: string
): string {
  const nextState = getNextState(path, event, ctx);

  req.session.user.journey = {
    nextPath: nextState.value,
    optionalPaths:
      Object.keys(nextState.meta).length > 0
        ? nextState.meta["AUTH." + nextState.value].optionalPaths
        : [],
  };

  req.log.info(
    `User journey transitioned from ${req.path} to ${nextState.value} with session id ${sessionId}`
  );

  if (!nextState) {
    throw Error(
      `Invalid user journey. No transition found from ${path} with event ${event}`
    );
  }

  return nextState.value;
}

export const JOURNEY_TYPE = {
  SIGN_IN: "sign-in",
  CREATE_ACCOUNT: "create-account",
};
