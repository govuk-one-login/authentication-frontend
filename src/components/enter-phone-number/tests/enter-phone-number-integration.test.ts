import request from "supertest";
import { describe } from "mocha";
import { expect, sinon } from "../../../../test/utils/test-utils";
import cheerio from "cheerio";
import decache from "decache";
import { HTTP_STATUS_CODES, PATH_NAMES } from "../../../app.constants";
import { ERROR_CODES, pathWithQueryParam } from "../../common/constants";
import nock = require("nock");

describe("Integration::enter phone number", () => {
  let token: string | string[];
  let cookies: string;
  let app: any;
  let baseApi: string;

  before(async () => {
    decache("../../../app");
    decache("../../../middleware/session-middleware");
    const sessionMiddleware = require("../../../middleware/session-middleware");

    process.env.SUPPORT_2HR_LOCKOUT = "1";

    sinon
      .stub(sessionMiddleware, "validateSessionMiddleware")
      .callsFake(function (req: any, res: any, next: any): void {
        res.locals.sessionId = "tDy103saszhcxbQq0-mjdzU854";
        req.session.user = {
          email: "test@test.com",
          phoneNumber: "7867",
          journey: {
            nextPath: PATH_NAMES.CREATE_ACCOUNT_ENTER_PHONE_NUMBER,
          },
          isAccountCreationJourney: true,
        };

        next();
      });

    app = await require("../../../app").createApp();
    baseApi = process.env.FRONTEND_API_BASE_URL;

    request(app)
      .get(PATH_NAMES.CREATE_ACCOUNT_ENTER_PHONE_NUMBER)
      .end((err, res) => {
        const $ = cheerio.load(res.text);
        token = $("[name=_csrf]").val();
        cookies = res.headers["set-cookie"];
      });
  });

  beforeEach(() => {
    nock.cleanAll();
  });

  after(() => {
    process.env.SUPPORT_2HR_LOCKOUT = "0";
    sinon.restore();
  });

  it("should return enter phone number page", (done) => {
    request(app)
      .get(PATH_NAMES.CREATE_ACCOUNT_ENTER_PHONE_NUMBER)
      .expect(200, done);
  });

  it("should return error when csrf not present", (done) => {
    request(app)
      .post(PATH_NAMES.CREATE_ACCOUNT_ENTER_PHONE_NUMBER)
      .type("form")
      .send({
        phoneNumber: "123456789",
      })
      .expect(500, done);
  });

  it("should return validation error when uk phone number not entered", (done) => {
    request(app)
      .post(PATH_NAMES.CREATE_ACCOUNT_ENTER_PHONE_NUMBER)
      .type("form")
      .set("Cookie", cookies)
      .send({
        _csrf: token,
        phoneNumber: "",
      })
      .expect(function (res) {
        const $ = cheerio.load(res.text);
        expect($("#phoneNumber-error").text()).to.contains(
          "Enter a UK mobile phone number"
        );
      })
      .expect(400, done);
  });

  it("should return validation error when uk phone number entered is not valid", (done) => {
    request(app)
      .post(PATH_NAMES.CREATE_ACCOUNT_ENTER_PHONE_NUMBER)
      .type("form")
      .set("Cookie", cookies)
      .send({
        _csrf: token,
        phoneNumber: "123456789",
      })
      .expect(function (res) {
        const $ = cheerio.load(res.text);
        expect($("#phoneNumber-error").text()).to.contains(
          "Enter a UK mobile phone number"
        );
      })
      .expect(400, done);
  });

  it("should return validation error when uk phone number entered contains text", (done) => {
    request(app)
      .post(PATH_NAMES.CREATE_ACCOUNT_ENTER_PHONE_NUMBER)
      .type("form")
      .set("Cookie", cookies)
      .send({
        _csrf: token,
        phoneNumber: "123456789dd",
      })
      .expect(function (res) {
        const $ = cheerio.load(res.text);
        expect($("#phoneNumber-error").text()).to.contains(
          "Enter a UK mobile phone number using only numbers or the + symbol"
        );
      })
      .expect(400, done);
  });

  it("should return validation error when uk phone number entered less than 12 characters", (done) => {
    request(app)
      .post(PATH_NAMES.CREATE_ACCOUNT_ENTER_PHONE_NUMBER)
      .type("form")
      .set("Cookie", cookies)
      .send({
        _csrf: token,
        phoneNumber: "123",
      })
      .expect(function (res) {
        const $ = cheerio.load(res.text);
        expect($("#phoneNumber-error").text()).to.contains(
          "Enter a UK mobile phone number, like 07700 900000"
        );
      })
      .expect(400, done);
  });

  it("should return validation error when uk phone number entered greater than 12 characters", (done) => {
    request(app)
      .post(PATH_NAMES.CREATE_ACCOUNT_ENTER_PHONE_NUMBER)
      .type("form")
      .set("Cookie", cookies)
      .send({
        _csrf: token,
        phoneNumber: "123123123123123123",
      })
      .expect(function (res) {
        const $ = cheerio.load(res.text);
        expect($("#phoneNumber-error").text()).to.contains(
          "Enter a UK mobile phone number, like 07700 900000"
        );
      })
      .expect(400, done);
  });

  it("should redirect to /check-your-phone page when valid UK phone number entered", (done) => {
    nock(baseApi)
      .post("/send-notification")
      .once()
      .reply(HTTP_STATUS_CODES.NO_CONTENT);

    request(app)
      .post(PATH_NAMES.CREATE_ACCOUNT_ENTER_PHONE_NUMBER)
      .type("form")
      .set("Cookie", cookies)
      .send({
        _csrf: token,
        phoneNumber: "07738394991",
      })
      .expect("Location", PATH_NAMES.CHECK_YOUR_PHONE)
      .expect(302, done);
  });

  it("should return validation error when international phone number not entered", (done) => {
    request(app)
      .post(PATH_NAMES.CREATE_ACCOUNT_ENTER_PHONE_NUMBER)
      .type("form")
      .set("Cookie", cookies)
      .send({
        _csrf: token,
        hasInternationalPhoneNumber: true,
        internationalPhoneNumber: "",
        supportInternationalNumbers: true,
      })
      .expect(function (res) {
        const $ = cheerio.load(res.text);
        expect($("#internationalPhoneNumber-error").text()).to.contains(
          "Enter a mobile phone number"
        );
        expect($("#phoneNumber-error").text()).to.contains("");
      })
      .expect(400, done);
  });

  it("should return validation error when international phone number entered is not valid", (done) => {
    request(app)
      .post(PATH_NAMES.CREATE_ACCOUNT_ENTER_PHONE_NUMBER)
      .type("form")
      .set("Cookie", cookies)
      .send({
        _csrf: token,
        hasInternationalPhoneNumber: true,
        internationalPhoneNumber: "123456789",
        supportInternationalNumbers: true,
      })
      .expect(function (res) {
        const $ = cheerio.load(res.text);
        expect($("#internationalPhoneNumber-error").text()).to.contains(
          "Enter a mobile phone number in the correct format, including the country code"
        );
        expect($("#phoneNumber-error").text()).to.contains("");
      })
      .expect(400, done);
  });

  it("should return validation error when international phone number entered contains text", (done) => {
    request(app)
      .post(PATH_NAMES.CREATE_ACCOUNT_ENTER_PHONE_NUMBER)
      .type("form")
      .set("Cookie", cookies)
      .send({
        _csrf: token,
        hasInternationalPhoneNumber: true,
        internationalPhoneNumber: "123456789dd",
        supportInternationalNumbers: true,
      })
      .expect(function (res) {
        const $ = cheerio.load(res.text);
        expect($("#internationalPhoneNumber-error").text()).to.contains(
          "Enter a mobile phone number using only numbers or the + symbol"
        );
        expect($("#phoneNumber-error").text()).to.contains("");
      })
      .expect(400, done);
  });

  it("should return validation error when international phone number entered less than 8 characters", (done) => {
    request(app)
      .post(PATH_NAMES.CREATE_ACCOUNT_ENTER_PHONE_NUMBER)
      .type("form")
      .set("Cookie", cookies)
      .send({
        _csrf: token,
        hasInternationalPhoneNumber: true,
        internationalPhoneNumber: "1234567",
        supportInternationalNumbers: true,
      })
      .expect(function (res) {
        const $ = cheerio.load(res.text);
        expect($("#internationalPhoneNumber-error").text()).to.contains(
          "Enter a mobile phone number in the correct format, including the country code"
        );
        expect($("#phoneNumber-error").text()).to.contains("");
      })
      .expect(400, done);
  });

  it("should return validation error when international phone number entered greater than 16 characters", (done) => {
    request(app)
      .post(PATH_NAMES.CREATE_ACCOUNT_ENTER_PHONE_NUMBER)
      .type("form")
      .set("Cookie", cookies)
      .send({
        _csrf: token,
        hasInternationalPhoneNumber: true,
        internationalPhoneNumber: "12345678901234567",
        supportInternationalNumbers: true,
      })
      .expect(function (res) {
        const $ = cheerio.load(res.text);
        expect($("#internationalPhoneNumber-error").text()).to.contains(
          "Enter a mobile phone number in the correct format, including the country code"
        );
        expect($("#phoneNumber-error").text()).to.contains("");
      })
      .expect(400, done);
  });

  it("should redirect to /check-your-phone page when valid international phone number entered", (done) => {
    nock(baseApi)
      .post("/send-notification")
      .once()
      .reply(HTTP_STATUS_CODES.NO_CONTENT);

    request(app)
      .post(PATH_NAMES.CREATE_ACCOUNT_ENTER_PHONE_NUMBER)
      .type("form")
      .set("Cookie", cookies)
      .send({
        _csrf: token,
        hasInternationalPhoneNumber: true,
        internationalPhoneNumber: "+33645453322",
        supportInternationalNumbers: true,
      })
      .expect("Location", PATH_NAMES.CHECK_YOUR_PHONE)
      .expect(302, done);
  });

  it('should render 2hr lockout "You asked for too many codes" error page when request OTP more than 5 times', (done) => {
    nock(baseApi).persist().post("/send-notification").times(6).reply(400, {
      code: ERROR_CODES.VERIFY_PHONE_NUMBER_MAX_CODES_SENT,
      success: false,
    });

    request(app)
      .post(PATH_NAMES.CREATE_ACCOUNT_ENTER_PHONE_NUMBER)
      .type("form")
      .set("Cookie", cookies)
      .send({
        _csrf: token,
        phoneNumber: "07738394991",
      })
      .expect((res) => {
        res.text.includes(
          "You asked to resend the security code too many times"
        );
      })
      .expect((res) => {
        res.text.includes("You will not be able to continue for 2 hours.");
      })
      .expect(200, done);
  });

  it('should redirect to SECURITY_CODE_REQUEST_EXCEEDED and render the 2hr lockout "you cannot create" error page when user has exceeded the OTP request limit', (done) => {
    nock(baseApi).post("/send-notification").once().reply(400, {
      code: ERROR_CODES.VERIFY_PHONE_NUMBER_CODE_REQUEST_BLOCKED,
      success: false,
    });

    request(app)
      .post(PATH_NAMES.CREATE_ACCOUNT_ENTER_PHONE_NUMBER)
      .type("form")
      .set("Cookie", cookies)
      .send({
        _csrf: token,
        phoneNumber: "07738394991",
      })
      .expect((res) => {
        res.text.includes("You cannot create a GOV.UK One Login");
      })
      .expect((res) => {
        res.text.includes("Wait 2 hours");
      })
      .expect(
        "Location",
        pathWithQueryParam(
          PATH_NAMES.SECURITY_CODE_REQUEST_EXCEEDED,
          "actionType",
          "otpBlocked"
        )
      )
      .expect(302, done);
  });
});
