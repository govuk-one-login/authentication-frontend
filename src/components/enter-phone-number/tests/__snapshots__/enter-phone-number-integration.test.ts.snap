// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration::enter phone number should render 2hr lockout "You asked for too many codes" error page when request OTP more than 5 times 1`] = `
Object {
  "contentId": "",
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "create account",
  "taxonomyLevel3": "",
  "taxonomyLevel4": "",
  "taxonomyLevel5": "",
}
`;

exports[`Integration::enter phone number should return enter phone number page 1`] = `
Object {
  "contentId": "0f519eb6-5cd4-476f-968f-d847b3c4c034",
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "create account",
  "taxonomyLevel3": "",
  "taxonomyLevel4": "",
  "taxonomyLevel5": "",
}
`;

exports[`Integration::enter phone number should return error when csrf not present 1`] = `
Object {
  "contentId": undefined,
  "taxonomyLevel1": undefined,
  "taxonomyLevel2": undefined,
  "taxonomyLevel3": undefined,
  "taxonomyLevel4": undefined,
  "taxonomyLevel5": undefined,
}
`;

exports[`Integration::enter phone number should return validation error when international phone number entered contains text 1`] = `
Object {
  "contentId": "",
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "create account",
  "taxonomyLevel3": "",
  "taxonomyLevel4": "",
  "taxonomyLevel5": "",
}
`;

exports[`Integration::enter phone number should return validation error when international phone number entered greater than 16 characters 1`] = `
Object {
  "contentId": "",
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "create account",
  "taxonomyLevel3": "",
  "taxonomyLevel4": "",
  "taxonomyLevel5": "",
}
`;

exports[`Integration::enter phone number should return validation error when international phone number entered is not valid 1`] = `
Object {
  "contentId": "",
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "create account",
  "taxonomyLevel3": "",
  "taxonomyLevel4": "",
  "taxonomyLevel5": "",
}
`;

exports[`Integration::enter phone number should return validation error when international phone number entered less than 8 characters 1`] = `
Object {
  "contentId": "",
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "create account",
  "taxonomyLevel3": "",
  "taxonomyLevel4": "",
  "taxonomyLevel5": "",
}
`;

exports[`Integration::enter phone number should return validation error when international phone number not entered 1`] = `
Object {
  "contentId": "",
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "create account",
  "taxonomyLevel3": "",
  "taxonomyLevel4": "",
  "taxonomyLevel5": "",
}
`;

exports[`Integration::enter phone number should return validation error when uk phone number entered contains text 1`] = `
Object {
  "contentId": "",
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "create account",
  "taxonomyLevel3": "",
  "taxonomyLevel4": "",
  "taxonomyLevel5": "",
}
`;

exports[`Integration::enter phone number should return validation error when uk phone number entered greater than 12 characters 1`] = `
Object {
  "contentId": "",
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "create account",
  "taxonomyLevel3": "",
  "taxonomyLevel4": "",
  "taxonomyLevel5": "",
}
`;

exports[`Integration::enter phone number should return validation error when uk phone number entered is not valid 1`] = `
Object {
  "contentId": "",
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "create account",
  "taxonomyLevel3": "",
  "taxonomyLevel4": "",
  "taxonomyLevel5": "",
}
`;

exports[`Integration::enter phone number should return validation error when uk phone number entered less than 12 characters 1`] = `
Object {
  "contentId": "",
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "create account",
  "taxonomyLevel3": "",
  "taxonomyLevel4": "",
  "taxonomyLevel5": "",
}
`;

exports[`Integration::enter phone number should return validation error when uk phone number not entered 1`] = `
Object {
  "contentId": "",
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "create account",
  "taxonomyLevel3": "",
  "taxonomyLevel4": "",
  "taxonomyLevel5": "",
}
`;
